working/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,382 @@
+encapsulation.
+
+## Proposal
+
+The basic idea is simple. We let users use a private name in a named parameter.
+The compiler removes the `_` from the argument name but keeps it for the
+corresponding initialized or declared field. In other words, we do exactly what
+users are doing by hand when they write an initializer like:
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({
+    int? windows,
+    int? bedrooms,
+    int? swimmingPools,
+  })  : _windows = windows,
+        _bedrooms = bedrooms,
+        _swimmingPools = swimmingPools;
+}
+```
+
+With this proposal, the above class can be written with identical semantics like
+so:
+
+```dart
+class House {
+  int? _windows;
+  int? _bedrooms;
+  int? _swimmingPools;
+
+  House({this._windows, this._bedrooms, this._swimmingPools});
+}
+```
+
+When this proposal is combined with primary constructors, they can write:
+
+```dart
+class House({
+  int? _windows,
+  int? _bedrooms,
+  int? _swimmingPools,
+});
+```
+
+## Static semantics
+
+An identifier is a **private name** if it starts with an underscore (`_`),
+otherwise it's a **public name**.
+
+A private name may have a **corresponding public name**. If the characters of
+the identifier with the leading underscore removed form a valid identifier and a
+public name, then that is the private name's corresponding public name. *For
+example, the corresponding public name of `_foo` is `foo`.* If removing the
+underscore does not leave something which is is a valid identifier *(as in `_`
+or `_2x`)* or leaves another private name *(as in `__x`)*, then the private name
+has no corresponding public name.
+
+Given an initializing formal or field parameter (for a primary constructor) with
+private name *p* in constructor C:
+
+*   If *p* has no corresponding public name *n*, then compile-time error. *You
+    can't use a private name for a named parameter unless there is a valid
+    public name that could be used at the callsite.*
+
+*   If any other parameter in C has declared name *p* or *n*, then
+    compile-time error. *If removing the `_` leads to a collision with
+    another parameter, then there is a conflict.*
+
+*   Otherwise, the name of the parameter in C is *n*. *If the parameter is
+    named, this then avoids the compile-time error that would otherwise be
+    reported for a private named parameter.*
+
+*   The local variable in the initializer list scope of C is *p*. *In the
+    initializer list, the private name is used. Inside the body of the
+    constructor, uses of *p* refer to the field, not the parameter.*
+
+*   If the parameter is an initializing formal, then it initializes a
+    corresponding field with name *p*.
+
+*   Else the field parameter induces an instance field with name *p*.
+
+*For example:*
+
+```dart
+class Id {
+  late final int _region = 0;
+
+  this({this._region, final int _value}) : assert(_region > 0 && _value > 0);
```

Comment:
Even here, I think it would be helpful to add a comment saying something like `// If primary constructors are supported.`